Source/CombatExtended/Harmony/Harmony_GenRadial.cs

```diff
@@ -79,54 +79,92 @@ internal static void Patch()
-                {
-                    max_count = mid_count;
-                }
+                int mid = (lowerBound + upperBound) >> 1;
+                if (RadialPatternRadii[mid] <= radius)
+                    lowerBound = mid;
                 else
-                {
-                    count = mid_count;
-                }
+                    upperBound = mid;
             }
 #endif
-            float start = radialPatternRadii[count];
-            while (start <= radius)
-            {
-                start = radialPatternRadii[count++];
-            }
-            __result = count - 1;
+            // Linear search:
+            // When (int)radius + (int)diagonal is odd,  the answer is always 8n + 5
+            // When (int)radius + (int)diagonal is even, the answer is always 8n + 1
+            // So we can set the last three bits of lowerBound
+            // and check only every 8 numbers
+            lowerBound = (lowerBound & -7) | (HasOddCells(radius) ? 5 : 1);
+            while (RadialPatternRadii[lowerBound] <= radius)
+                lowerBound += 8;
+            __result = lowerBound;
             return false;
-
+        }
+        
+        // A helper function for Prefix_NumCellsInRadius()
+        // Evaluates if (int)r + (int)d is an even number (round toward zero)
+        // Where d = r * sqrt(1/2)
+        // When (int)radius + (int)diagonal is odd,  the answer is always 8n + 5
+        // When (int)radius + (int)diagonal is even, the answer is always 8n + 1
+        // Use IEEE-754 bit hacks
+        //[MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool HasOddCells(float r)
+        {
+            const float SqrtHalf = 0.707106781f;
+            float d = r * SqrtHalf;
+            // bit hack magic
+            int bitr = BitConverter.SingleToInt32Bits(r);
+            int bitd = BitConverter.SingleToInt32Bits(d);
+            // Slightly different from the real IEEE 754 mantissa
+            //   -- we first take the fractional part
+            //   -- then add back the missing starting 1
+            int mantissar = bitr & 0x007F_FFFF | 0x0080_0000;
+            int mantissad = bitd & 0x007F_FFFF | 0x0080_0000;
+            // We want to shift the fractional part out
+            //   -- shift 23, plus or minus the exponent part
+            //   -- the exponenet part itself is offset by 127
+            int shiftr = 23 + 127 - ((bitr >> 23) & 0xFF);
+            int shiftd = 23 + 127 - ((bitd >> 23) & 0xFF);
+            return (((mantissar >> shiftr) ^ (mantissad >> shiftd)) & 1) != 0;
+        }
+        
+        // A separate method to handle when radius exceeds MAX_RADIUS
+        // Apparently loading all these strings for formatting is very slow
+        // Even when they are not executed, likely due to branch predictions
+        // Separating the rare & slow stuff to this method helps a lot
+        private static void CheckRadius(float radius)
```

Comment:
A bit of bikeshedding, but the only way execution reaches this point is if the conditional is `true`, so it will *always* result in the error message. So best to simply have it unconditionally throw an error, with a name to match.  Moving the string formatting out is a good idea though.

Reply:
Technically, the check inside the function was a strict "radius > MAX_RADIUS" while the one outside was checking for "radius >= MAX_RADIUS" which also allows for equals.
But I have moved the check into the main function for better readability, I've also renamed the Error Log method accordingly.